src/sinks/socket.rs

```diff
@@ -164,63 +159,97 @@ mod test {
-                // ran into the issue before!
-                tls: None,
+                tls: Some(TlsConfig {
+                    enabled: Some(true),
+                    options: TlsOptions {
+                        verify_certificate: Some(false),
+                        verify_hostname: Some(false),
+                        ca_path: Some("tests/data/localhost.crt".into()),
+                        ..Default::default()
+                    },
+                }),
             }),
         };
         let mut rt = runtime();
         let context = SinkContext::new_test(rt.executor());
         let (sink, _healthcheck) = config.build(context).unwrap();
 
-        let close = Arc::new(tokio::sync::Notify::new());
-        let counter = Arc::new(AtomicUsize::new(0));
-
-        let close1 = close.clone();
-        let counter1 = counter.clone();
-        let jh = rt.spawn_handle(async move {
-            let mut listener = TcpListener::bind(&addr).await.unwrap();
+        let msg_counter = Arc::new(AtomicUsize::new(0));
+        let msg_counter1 = msg_counter.clone();
+        let conn_counter = Arc::new(AtomicUsize::new(0));
+        let conn_counter1 = conn_counter.clone();
 
-            // Only accept two connections after the second connection is done
-            // we can exit.
-            for _ in 0..2 {
-                let (mut conn, _) = listener.accept().await.unwrap();
+        let (mut close_tx, mut close_rx) = tokio01::sync::mpsc::unbounded_channel::<()>();
 
+        let config = Some(TlsConfig {
+            enabled: Some(true),
+            options: TlsOptions {
+                crt_path: Some("tests/data/localhost.crt".into()),
+                key_path: Some("tests/data/localhost.key".into()),
+                ..Default::default()
+            },
+        });
+        let stream = MaybeTlsSettings::from_config(&config, true).unwrap();
+
+        // Only accept two connections.
+        let fut = stream
+            .bind(&addr)
+            .unwrap()
+            .incoming()
+            .take(2)
+            .for_each(move |mut socket| {
+                conn_counter1.fetch_add(1, Ordering::Relaxed);
                 loop {
-                    let mut buf = vec![0u8; 11];
+                    if let Ok(Async::Ready(_)) = close_rx.poll() {
+                        socket.shutdown().unwrap();
+                        break;
+                    }
 
-                    tokio::select! {
-                        _ = close1.notified() => {
-                            conn.shutdown(std::net::Shutdown::Write).unwrap();
-                            break;
+                    let mut buf = vec![0u8; 11];
+                    match socket.read(&mut buf) {
+                        Err(error) if error.kind() == ErrorKind::WouldBlock => {}
+                        Err(error) => {
+                            // How compare crate::tls::TlsError::HandshakeNotReady with `error.into_inner()`?
```

Comment:
You can use this downcast functionality to get the original error https://doc.rust-lang.org/stable/std/error/trait.Error.html#method.downcast-2

Reply:
Thanks for the hint!
While I tried downcast error I realised that `HandshakeNotReady` will happened only when error kind is `WouldBlock`, so I removed this extra if-branch where condition will be always evaluated to `true`.